### PR TITLE
Tools: add a --speedup parameter to autotest to control speed of all …

### PR DESCRIPTION
**Pull request by peterbarker**
…sims

### CODE REVIEW

File: Tools/autotest/arducopter.py [hunk 1]
```diff
@@ -972,7 +971,7 @@ def setup_rc(mavproxy):
     mavproxy.send('rc 3 1000\n')
 
 
-def fly_ArduCopter(binary, viewerip=None, use_map=False, valgrind=False, gdb=False, frame=None, params=None, gdbserver=False):
+def fly_ArduCopter(binary, viewerip=None, use_map=False, valgrind=False, gdb=False, frame=None, params=None, gdbserver=False, speedup=30):
```
Comment (WickedShell) [inline Tools/autotest/arducopter.py]: 30x is to fast, 20x is stable though.

**Comment by khancyr** [comment]
The only problem with that is that all timed failsafe are broken since mavproxy lag at 1Hz heartbeat.

**Comment by WickedShell** [comment]
@khancyr This should be a no change. We already ran all of the SITL tests at 10x, so this isn't a change in behavior, it's highly possible it was already broken, but we were running at 10x already.

**Comment by peterbarker** [comment]
On Sun, 27 Aug 2017, WickedShell wrote:

> @khancyr This should be a no change. We already ran all of the SITL tests at 10x, so this isn't a change in behavior, it's highly possible it was already broken, but we were running at 10x already.

Should mention that running at higher rates seems to fail on RC overrides. 
I theorise that we're not sending packets to ArduPilot fast enough for 
continual overriding of the input.

We probably have to have MAVProxy adapt itself to the higher clock rates.

File: Tools/autotest/autotest.py [hunk 2]
```diff
@@ -444,6 +444,7 @@ def run_tests(steps):
     parser.add_option("--map", action='store_true', default=False, help='show map')
     parser.add_option("--experimental", default=False, action='store_true', help='enable experimental tests')
     parser.add_option("--timeout", default=3000, type='int', help='maximum runtime in seconds')
+    parser.add_option("--speedup", default=10, type='int', help='speedup to runthe simulations at')
```
Comment (OXINARF) [inline Tools/autotest/autotest.py]: runthe -> run the

I don't have a better suggestion, but having a default here will override the defaults in vehicles.

File: Tools/autotest/arducopter.py [hunk 3]
```diff
@@ -972,7 +971,7 @@ def setup_rc(mavproxy):
     mavproxy.send('rc 3 1000\n')
 
 
-def fly_ArduCopter(binary, viewerip=None, use_map=False, valgrind=False, gdb=False, frame=None, params=None, gdbserver=False):
+def fly_ArduCopter(binary, viewerip=None, use_map=False, valgrind=False, gdb=False, frame=None, params=None, gdbserver=False, speedup=20):
```
Comment (OXINARF) [inline Tools/autotest/arducopter.py]: I know @WickedShell is looking at increasing this to 20, but I'd prefer that change in a separate commit. Also, as I wrote in another comment, this default value won't actually matter.

File: Tools/autotest/arducopter.py [hunk 4]
```diff
@@ -1329,15 +1328,15 @@ def fly_ArduCopter(binary, viewerip=None, use_map=False, valgrind=False, gdb=Fal
     return True
 
 
-def fly_CopterAVC(binary, viewerip=None, use_map=False, valgrind=False, gdb=False, frame=None, params=None, gdbserver=False):
+def fly_CopterAVC(binary, viewerip=None, use_map=False, valgrind=False, gdb=False, frame=None, params=None, gdbserver=False, speedup=30):
```
Comment (OXINARF) [inline Tools/autotest/arducopter.py]: Any reason this was changed to a default of 30?

File: Tools/autotest/autotest.py (hunk 2 above)
Comment (WickedShell) [reply Tools/autotest/autotest.py]: I'm fine with the override if your lazy as long as nothing breaks. (Mostly cause I didn't have a better version, and didn't think it was important enough to focus on) As far as the CI side goes we call autotest for each target so we can pass different values there as needed to speed it up.

File: Tools/autotest/arducopter.py (hunk 3 above)
Comment (WickedShell) [reply Tools/autotest/arducopter.py]: Leaving the 20x to the log download one is fine with me, I can reintroduce it then.

**Comment by peterbarker** [comment]
On Tue, 29 Aug 2017, WickedShell wrote:

> I'm fine with the override if your lazy as long as nothing breaks. (Mostly
> cause I didn't have a better version, and didn't think it was important
> enough to focus on) As far as the CI side goes we call autotest for each
> target so we can pass different values there as needed to speed it up.

I've come up with something which won't override the vehicle defaults 
unless speedup is passed as a parameter.  Will push soon.

File: Tools/autotest/arducopter.py (hunk 3 above)
Comment (peterbarker) [reply Tools/autotest/arducopter.py]: @OXINARF Good call, it's not stable for me @ 20 - and I didn't actually mean for this to slip in!

File: Tools/autotest/arducopter.py (hunk 4 above)
Comment (peterbarker) [reply Tools/autotest/arducopter.py]: I was playing around with mavnative the other night and this crept in...  reverted to 10

**Comment by peterbarker** [comment]
@OXINARF @WickedShell new version pushed up

**Comment by OXINARF** [comment]
@WickedShell I think that's possible, @peterbarker certainly knows better than I do.

@peterbarker Merged, thanks!

**Comment by peterbarker** [comment]
You can choose not to unpack the arguments and slurp them all into a `**kwargs` thing - you'd then have to unpack them yourself later if you needed *some* of the arguments.

Not a great fan, frankly.  Passing unnamed, untyped opaque blobs down isn't a great look IMO - there are better ways.  So perhaps arguments for SITL could be put into a start_SITL_args object or something, so at least they're all collected together...

That opinion's coming from someone who passed around `optionsref`s like they were candy in Perl, BTW.  Kind of an idium where the last argument to a function is a reference to a hash.  Functions adding/overriding often did `$otherobj->foo(arg1, arg, {%$optionsref, mybar => 1})`.

**Pull request merged**